File: Kawai codes/05.ODE1/ch05pr12.m
%**************************************************************************
%*     Section 5.5.6                                                      *
%*     filename: ch05pr12.m                                               *
%*     program listing number: 5.12                                       *
%*                                                                        *
%*     This program calculate the trajectory of a particle scattered by   *
%*     Yukawa potential using he Valet method.                            *
%*                                                                        *
%*     Programed by Ryoichi Kawai for Computational Physics Course        *
%*     Revised on 01/07/2014.                                             *
%**************************************************************************
clear all;

E=input('Enter Energy =');

% control parameter
bmax=3;
N=10;
db=bmax/N;
h=0.01;

subplot(1,2,1);

for i=1:2*N+1
    
    % set impact parameter
    b(i)=db*(i-N-1);
    
    % initial conditions
    x(1)=-10; y(1)=b(i); vx(1)=sqrt(2*E); vy(1)=0;
    
    % first Euler step
    r = sqrt(x(1)^2+y(1)^2);
    Fx=x(1)/r^2 * (1/r+1) * exp(-r);
    Fy=y(1)/r^2 * (1/r+1) * exp(-r);
    x(2)=x(1)+vx(1)*h+Fx*h^2/2;
    y(2)=y(1)+vy(1)*h+Fy*h^2/2;
    
    % Verlet method
    n=2;
    while abs(x(n))<10
        r = sqrt(x(n)^2+y(n)^2);
        Fx=x(n)/r^2 * (1/r+1) * exp(-r);
        Fy=y(n)/r^2 * (1/r+1) * exp(-r);
        x(n+1)=2*x(n)-x(n-1)+Fx*h^2;
        y(n+1)=2*y(n)-y(n-1)+Fy*h^2;
        n=n+1;
    end
    
    % final velocity
    vfx=(x(n)-x(n-2))/(2*h);
    vfy=(y(n)-y(n-2))/(2*h);
    
    % scattering angle
    theta(i) = acos((vx(1)*vfx+vy(1)*vfy)/(2*E));
    
    % plot the trajctory
    plot(x(1:n),y(1:n));
    hold on
end

% draw a target atom
p=plot(0,0,'o');
set(p,'Color','red','Linewidth',3);
hold off

axis([-10,10,-10,10]);
xlabel('x');
ylabel('y');

% plot scattering angle
subplot(1,2,2);
q=plot(b,theta);
xlabel('Impact Parameter');
ylabel('Scattering Angle theta');
set(q,'Linewidth',2);


 

File: Kawai codes/05.ODE1/ch05pr12.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
%**************************************************************************
%*     Section 5.5.6                                                      *
%*     filename: ch05pr12.py                                              *
%*     program listing number: 5.12                                       *
%*                                                                        *
%*     This program calculate the trajectory of a particle scattered by   *
%*     Yukawa potential using he Valet method.                            *
%*                                                                        *
%*     Programed by Ryoichi Kawai for Computational Physics Course        *
%*     Revised on 01/07/2014.                                             *
%**************************************************************************
"""
import numpy as np
import matplotlib.pyplot as plt


E=np.float(input('Enter Energy = '))

# control parameter
bmax=3
N=10
db=bmax/N
h=0.01
Kmax=10001
b=np.linspace(-bmax,bmax,2*N+1)
theta=np.zeros(2*N+1)
x=np.zeros(Kmax)
y=np.zeros(Kmax)
vx=np.zeros(Kmax)
vy=np.zeros(Kmax)
plt.ioff()
plt.figure(figsize=(12,5))

plt.subplot(1,2,1);

for i in range(0,2*N+1):
    
    # initial conditions
    x[0]=-10; y[0]=b[i]; vx[0]=np.sqrt(2.0*E); vy[0]=0.0;
    
    # first Euler step
    r = np.sqrt(x[0]**2+y[0]**2)
    Fx=x[0]/r**2 * (1.0/r+1.0) * np.exp(-r)
    Fy=y[0]/r**2 * (1.0/r+1.0) * np.exp(-r)
    x[1]=x[0]+vx[0]*h+Fx*h**2/2.0
    y[1]=y[0]+vy[0]*h+Fy*h**2/2.0
    
   # Verlet method
    n=1
    while r <11.0:
        r = np.sqrt(x[n]**2+y[n]**2)
        Fx=x[n]/r**2 * (1.0/r+1.0) * np.exp(-r)
        Fy=y[n]/r**2 * (1.0/r+1.0) * np.exp(-r)
        x[n+1]=2.0*x[n]-x[n-1]+Fx*h**2
        y[n+1]=2.0*y[n]-y[n-1]+Fy*h**2
        n+=1
    
    # final velocity
    vfx=(x[n]-x[n-2])/(2.0*h)
    vfy=(y[n]-y[n-2])/(2.0*h)
    
    # scattering angle
    theta[i] = np.arccos((vx[0]*vfx+vy[0]*vfy)/(2.0*E))
    
    # plot the trajctory
    plt.plot(x[0:n],y[0:n],'-k')


# draw a target atom
plt.plot(0,0,'or')

plt.xlabel('x');
plt.ylabel('y');

# plot scattering angle
plt.subplot(1,2,2)
plt.plot(b,theta)
plt.xlabel('Impact Parameter')
plt.ylabel('Scattering Angle theta')
plt.show()


 